File: dotnet/src/Workflow.cs
using Microsoft.Extensions.Logging;
using Temporalio.Common;
using Temporalio.Workflows;

namespace TemporalTripBooking;

[Workflow("BookWorkflow")]
public class BookWorkflow
{
    [WorkflowRun]
    public async Task<string> Execute(Shared input)
    {
        Workflow.Logger.LogInformation("Book workflow started, user_id = {}", input.UserId);
        var options = new ActivityOptions() { 
            StartToCloseTimeout = TimeSpan.FromSeconds(5),
            RetryPolicy = new() {
                InitialInterval = TimeSpan.FromSeconds(1),
                MaximumInterval = TimeSpan.FromSeconds(30),
                BackoffCoefficient = 2
            }
        };
        
        var compensations = new Stack<Func<Task>>();

        try
        {
            // Book Flight
            var flight = await Workflow.ExecuteActivityAsync(() =>
                TripActivities.BookFlight(input), options);
            compensations.Push(async () => 
                await Workflow.ExecuteActivityAsync(() => TripActivities.UndoBookFlight(input), options));
            
            Workflow.Logger.LogInformation("Sleeping for 1 second...");
            await Workflow.DelayAsync(TimeSpan.FromSeconds(1));

            // Book Hotel
            var hotel = await Workflow.ExecuteActivityAsync(() =>
                TripActivities.BookHotel(input), options);
            compensations.Push(async () => 
                await Workflow.ExecuteActivityAsync(() => TripActivities.UndoBookHotel(input), options));
            
            Workflow.Logger.LogInformation("Sleeping for 1 second...");
            await Workflow.DelayAsync(TimeSpan.FromSeconds(1));

            // Book Car
            var car = await Workflow.ExecuteActivityAsync(() =>
                TripActivities.BookCar(input), options);
            compensations.Push(async () => 
                await Workflow.ExecuteActivityAsync(() => TripActivities.UndoBookCar(input), options));

            // Notify User
            await Workflow.ExecuteActivityAsync(() =>
                TripActivities.NotifyUser(input), options);

            return $"{flight} {hotel} {car}";
        }
        catch (Exception ex)
        {
            Workflow.Logger.LogError(ex, "Failed to book trip");
            while (compensations.TryPop(out var compensation))
            {
                try
                {
                    await compensation();
                }
                catch (Exception compEx)
                {
                    Workflow.Logger.LogWarning(compEx, "Compensation failed");
                }
            }

            return "Booking cancelled";
        }
    }
}

File: ruby/src/workflow.rb
# frozen_string_literal: true

require_relative 'activities'
require_relative 'shared'
require 'temporalio/retry_policy'
require 'temporalio/workflow'

module TripBooking
  class BookWorkflow < Temporalio::Workflow::Definition
    def execute(input_hash)
      input = TripBooking::Models::BookTripInput.new(**input_hash)
      Temporalio::Workflow.logger.info("Book workflow started, userId = #{input.userId}")

      retry_policy = Temporalio::RetryPolicy.new(
        max_interval: 30.0
      )

      # Saga compensations
      compensations = []

      # Book flight
      flight = Temporalio::Workflow.execute_activity(
        Activities::BookFlight,
        input,
        start_to_close_timeout: 5.0,
        retry_policy: retry_policy
      )
      compensations << Activities::UndoBookFlight

      Temporalio::Workflow.logger.info('Sleeping for 1 second...')
      Temporalio::Workflow.sleep(1.0)

      # Book hotel
      hotel = Temporalio::Workflow.execute_activity(
        Activities::BookHotel,
        input,
        start_to_close_timeout: 5.0,
        retry_policy: retry_policy
      )
      compensations << Activities::UndoBookHotel

      Temporalio::Workflow.logger.info('Sleeping for 1 second...')
      Temporalio::Workflow.sleep(1.0)

      # Book car
      car = Temporalio::Workflow.execute_activity(
        Activities::BookCar,
        input,
        start_to_close_timeout: 5.0,
        retry_policy: retry_policy
      )
      compensations << Activities::UndoBookCar

      # Notify user
      Temporalio::Workflow.execute_activity(
        Activities::NotifyUser,
        input,
        start_to_close_timeout: 5.0,
        retry_policy: retry_policy
      )

      "#{flight}, #{hotel}, #{car}"
    rescue StandardError => e
      Temporalio::Workflow.logger.error("Failed to book trip #{e}")
      compensations.reverse_each do |compensation|
        Temporalio::Workflow.execute_activity(
          compensation,
          input,
          start_to_close_timeout: 5.0,
          retry_policy: retry_policy
        )
      end
      'Booking canceled'
    end
  end
end
